src/ref_cnt.rs

```diff
@@ -92,7 +92,7 @@ unsafe impl<T> RefCnt for Arc<T> {
         Arc::into_raw(me) as *mut T
     }
     fn as_ptr(me: &Arc<T>) -> *mut T {
-        me as &T as *const T as *mut T
+        Arc::as_ptr(me) as *mut T
```

Comment:
This is a nice find.

Nevertheless, the Arc::as_ptr was introduced in 1.45 and this crate still tries to support 1.31. Back then this was somewhat a permissive/gray area and there wasn't a better way to do it anyway, but I agree the way forward is to use the dedicated method.

Do you think we can find some kind of trick for fallback?

Reply:
Hmm, it's even more interesting on 1.31 than not compiling :-D :

```
warning: function cannot return without recursing
  --> src/ref_cnt.rs:94:5
   |
94 |     fn as_ptr(me: &Arc<T>) -> *mut T {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
95 |         Arc::as_ptr(me) as *mut T
   |         --------------- recursive call site
   |
   = note: #[warn(unconditional_recursion)] on by default
   = help: a `loop` may express intention better if this is on purpose
```

Reply:
Ah, I'm not sure this is possible without an MSRV bump then :/. as_ptr is implemented using private fields, and since it's #[repr(Rust)] we can't transmute to get at them.

I'll look into this again today and close this if I can't find a fix.

Reply:
Don't close it :-). You've discovered the cause, that was the harder part.

The options that I see are:
* Figure some way to use `as_ptr` where available and fall back into the current state where not. Either by some trick, or by some kind of build-script detection.
* Bump the version and make sure it doesn't compile (instead of creating an infinite recursion).

Reply:
I'm looking into ways to prevent compilation if you choose to bump MSRV. The easiest way is `#[deny(unconditional_recursion)]` on the function with a comment explaining why in the code, but that isn't great for the user (having to open source code to understand the compile error). There may be some type inference hacks that can be done to give a more helpful error.